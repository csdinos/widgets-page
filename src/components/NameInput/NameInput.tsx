import {Button, FormControl, FormHelperText, Grid, TextField} from '@material-ui/core'
import {useEffect, useState} from 'react'
import {useHistory} from 'react-router-dom'
import {useDispatch, useSelector} from 'react-redux'
import {State} from '../../types'
import * as actions from '../../actions'

export const NameInput = () => {
  const history = useHistory()
  const dispatch = useDispatch()
  const newWidget = useSelector((state: State) => state.newWidget)
  const name = useSelector((state: State) => state.newWidget.name)
  const [error, setError] = useState(false)
  const [errorMessage, setErrorMessage] = useState('')

  useEffect(() => {
    if (!newWidget.language) {
      history.push('/')
    }
  })
  const onChange = (event: any) => dispatch(actions.setNewWidgetName(event.target.value))

  const validate = (): boolean => {
    if (!name?.length) {
      setError(true)
      setErrorMessage('Cannot be empty')

      return false
    }

    return true
  }

  const onCreate = () => {
    if (validate()) {
      newWidget.id = new Date().getTime() // surely this could have been handled better :D

      dispatch(actions.addWidget([newWidget]))
      dispatch(actions.setShouldUpdateStorage(true))

      history.push('/')
    }
  }
  return (
    <Grid container
          direction="row"
          alignItems="center"
          justify="center"
          spacing={2}
    >
      <Grid item xs={12}>
        <FormControl variant="outlined" fullWidth error={error}>
          <TextField
            id="name-text-input"
            label="Widget Name"
            variant="outlined"
            onChange={onChange}
            error={error}
          />
          <FormHelperText id="error-message" hidden={!error}>{errorMessage}</FormHelperText>
        </FormControl>
      </Grid>
      <Grid item xs={12}>
        <Button
          id="create-btn"
          variant="contained"
          color="primary"
          onClick={onCreate}
          fullWidth
        >
          Create
        </Button>
      </Grid>
    </Grid>
  )
}