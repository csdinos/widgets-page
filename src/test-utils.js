import React from 'react'
import {render as rtlRender} from '@testing-library/react'
import {createStore} from 'redux'
import {Provider} from 'react-redux'
import {reducer} from './reducers'

function render(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({children}) {
    return <Provider store={store}>{children}</Provider>
  }

  return rtlRender(ui, {wrapper: Wrapper, ...renderOptions})
}

export const initialState = {
  widgets: [],
  newWidget: {
    id: -1,
    name: '',
    language: '',
  },
  deleteModalId: null,
  hasLoaded: false,
  shouldUpdateStorage: false
}

export * from '@testing-library/react'
export {render}