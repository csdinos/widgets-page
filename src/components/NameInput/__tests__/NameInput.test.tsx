import React from "react";
import { render } from "../../../test-utils";
import {NameInput} from "../NameInput";

describe("<NameInput />", () => {
  test("should display an empty name input with a create button", async () => {
    const { asFragment } = render(<NameInput />);
    expect(asFragment()).toMatchSnapshot()
  });
});