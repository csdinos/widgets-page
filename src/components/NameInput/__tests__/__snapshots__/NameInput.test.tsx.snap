// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NameInput /> should display an empty name input with a create button 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="name-text-input"
              id="name-text-input-label"
            >
              Widget Name
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="name-text-input"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-3"
                >
                  <span>
                    Widget Name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-contained"
            hidden=""
            id="error-message"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
          id="create-btn"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Create
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
